internal/service/lakeformation/opt_in_test.go

```diff
@@ -0,0 +1,398 @@
+	rName := sdkacctest.RandomWithPrefix("tf-acc-test")
+	roleName := "aws_iam_role.test"
+	databaseName := "aws_glue_catalog_database.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.LakeFormation)
+		},
+		ErrorCheck:               acctest.ErrorCheck(t, names.LakeFormationServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckOptInDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOptInConfig_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckOptInExists(ctx, resourceName, &optin),
+					resource.TestCheckResourceAttr(resourceName, "principal.#", "1"),
+					resource.TestCheckResourceAttrPair(resourceName, "principal.0.data_lake_principal_identifier", roleName, names.AttrARN),
+					resource.TestCheckResourceAttr(resourceName, "resource_data.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "resource_data.0.database.#", "1"),
+					resource.TestCheckResourceAttrPair(resourceName, "resource_data.0.database.0.name", databaseName, names.AttrName),
+				),
+			},
+		},
+	})
+}
+
+func TestAccLakeFormationOptIn_disappears(t *testing.T) {
+	ctx := acctest.Context(t)
+	if testing.Short() {
+		t.Skip("skipping long-running test in short mode")
+	}
+
+	var optin lakeformation.ListLakeFormationOptInsOutput
+	resourceName := "aws_lakeformation_opt_in.test"
+	rName := sdkacctest.RandomWithPrefix("tf-acc-test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.LakeFormation)
+		},
+		ErrorCheck:               acctest.ErrorCheck(t, names.LakeFormationServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckOptInDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOptInConfig_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckOptInExists(ctx, resourceName, &optin),
+					acctest.CheckFrameworkResourceDisappears(ctx, acctest.Provider, tflakeformation.ResourceOptIn, resourceName),
+				),
+				ExpectNonEmptyPlan: false,
+			},
+		},
+	})
+}
+
+func testAccCheckOptInDestroy(ctx context.Context) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).LakeFormationClient(ctx)
+
+		for _, rs := range s.RootModule().Resources {
+			if rs.Type != "aws_lakeformation_opt_in" {
+				continue
+			}
+
+			principalID := rs.Primary.Attributes["principal.0.data_lake_principal_identifier"]
+
+			in := &lakeformation.ListLakeFormationOptInsInput{
+				Principal: &awstypes.DataLakePrincipal{
+					DataLakePrincipalIdentifier: &principalID,
+				},
+				Resource: &awstypes.Resource{},
+			}
+
+			if v, ok := rs.Primary.Attributes["resource_data.0.catalog.0.id"]; ok && v != "" {
```

Comment:
The way this works is a bit of a mess and if-else is one way to handle it. Idiomatic Go tries to avoid else entirely, however. I propose either switching to a switch or a separate return-early function that takes the attributes and returns the resource.